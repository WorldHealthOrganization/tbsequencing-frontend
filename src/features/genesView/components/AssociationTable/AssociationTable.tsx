/** @jsxImportSource @emotion/react */

import React from 'react';
import { SerializedStyles } from '@emotion/react';
import AppPaper from '../../../../components/AppPaper';
import H3 from '../../../../components/typography/H3';
import { associationTableStyles } from './styles';
import AppButton from '../../../../components/AppButton';
import LoadingWrapper from '../../../../components/LoadingWrapper';

export interface IAssociationTableData {
  id: number;
  leftColValue: number | string;
  rightColValue: number | string;
}

interface IAssociationTableProps {
  leftColumnItemComponent?: React.FC<{
    item: IAssociationTableData, onClick?: (value: IAssociationTableData) => void
  }>;
  tableHeader?: string;
  data?: IAssociationTableData[];
  onClick?: (value: IAssociationTableData) => void;
  isLoading: boolean;
  containerStyles?: SerializedStyles | {}
  noDataText?: string;
}

interface ILeftColumnButtonProps {
  item: IAssociationTableData,
  onClick?: (value: IAssociationTableData) => void;
}

export const LeftColumnButton = ({ item, onClick }: ILeftColumnButtonProps) => (
  <AppButton
    style={associationTableStyles.leftColumn}
    size="small"
    onClick={() => typeof onClick === 'function' && onClick(item)}
    variant="text"
    underlined
  >
    {item.leftColValue}
  </AppButton>
);

const DefaultLeftColumnItem = ({ text }: { text?: string }) => (
  <H3 style={associationTableStyles.defaultColumn}>{text}</H3>
);

const AssociationTable = ({
  tableHeader,
  data,
  leftColumnItemComponent,
  onClick,
  isLoading,
  containerStyles = {},
  noDataText = 'No data',
}: IAssociationTableProps) => {
  const LeftColumnItem = leftColumnItemComponent || DefaultLeftColumnItem;
  return (
    <AppPaper style={{
      ...associationTableStyles.paper,
      ...containerStyles,
    }}
    >
      <div>
        {tableHeader && <H3>{tableHeader}</H3>}
      </div>
      <LoadingWrapper isLoading={isLoading} centered>
        <div css={associationTableStyles.itemsContainer}>
          {!data?.length ? noDataText
            : data.map((item) => (
              <div key={item.id} css={associationTableStyles.singlePair}>
                <LeftColumnItem item={item} onClick={onClick} />
                <H3 style={associationTableStyles.rightColumnText}>
                  {item.rightColValue}
                </H3>
              </div>
            ))}
        </div>
      </LoadingWrapper>
    </AppPaper>
  );
};

export default AssociationTable;
