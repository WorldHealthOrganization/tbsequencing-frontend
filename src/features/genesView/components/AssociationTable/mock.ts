import { IGenesOverview, ISearchCount } from '../../models';

export const mostSearches: ISearchCount[] = [
  {
    counter: 112, geneName: 'rpoB', geneDbCrossrefId: 114, locusTag: 'aboba',
  },
  {
    counter: 97, geneName: 'rpoC', geneDbCrossrefId: 154, locusTag: 'aboba',
  },
  {
    counter: 90, geneName: 'Rv0669c', geneDbCrossrefId: 144, locusTag: '',
  },
  {
    counter: 50, geneName: 'J8670_RS10355', geneDbCrossrefId: 116, locusTag: '',
  },
  {
    counter: 22, geneName: 'J8670_RS10350', geneDbCrossrefId: 118, locusTag: '',
  },
  {
    counter: 18, geneName: 'cfp21', geneDbCrossrefId: 125, locusTag: '',
  },
  {
    counter: 12, geneName: 'J8670_RS10345', geneDbCrossrefId: 119, locusTag: '',
  },
  {
    counter: 9, geneName: 'vapB8', geneDbCrossrefId: 120, locusTag: '',
  },
  {
    counter: 3, geneName: 'vapC8', geneDbCrossrefId: 121, locusTag: '',
  },
];

export const simpleData: IGenesOverview = {
  endPos: 763325,
  id: 694,
  geneDbCrossrefId: 694,
  geneDescription: 'DNA-directed RNA polymerase subunit beta',
  geneName: 'rpoB',
  geneType: 'NP_215181.1',
  locusTag: 'Rv0667',
  ncbiId: '888164',
  proteinLength: 1172,
  startPos: 759807,
  strand: 1,
};

export const drugsTable = [
  { code: 'RIF', drugName: 'Rifampicine', drugId: 2 },
  { code: 'INH', drugName: 'Isoniazid', drugId: 1 },
  { code: 'EMB', drugName: 'Ethambutol', drugId: 4 },
  { code: 'LZD', drugName: 'Linezolid', drugId: 21 },
];

export const association = [
  { geneName: 'cfp21', drugs: [{ code: 'INH' }, { code: 'ZLD' }], geneDbCrossref: 1 },
  { geneName: 'J8670 RS10345', drugs: [{ code: 'INH' }, { code: 'ZLD' }], geneDbCrossref: 2 },
  { geneName: 'rpoB', drugs: [{ code: 'INH' }, { code: 'ZLD' }], geneDbCrossref: 694 },
  { geneName: 'Rv2490a', drugs: [{ code: 'INH' }, { code: 'ZLD' }], geneDbCrossref: 7 },
];
