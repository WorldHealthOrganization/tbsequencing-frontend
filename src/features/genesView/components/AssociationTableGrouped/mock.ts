export const association = [
  {
    geneName: 'cfp21', geneDbCrossrefId: 1525, drugId: 1, tier: 1,
  },
  {
    geneName: 'J8670 RS10345', geneDbCrossrefId: 1526, drugId: 1, tier: 1,
  },
  {
    geneName: 'rpoB', geneDbCrossrefId: 1527, drugId: 1, tier: 2,
  },
  {
    geneName: 'Rv2490a', geneDbCrossrefId: 1526, drugId: 1, tier: 2,
  },
];
