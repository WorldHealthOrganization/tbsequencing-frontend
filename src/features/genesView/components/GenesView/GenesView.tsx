/** @jsxImportSource @emotion/react */
import React, { SyntheticEvent, useEffect, useState } from 'react';
import { useSearchParams } from 'react-router-dom';
import BasePage from '../../../../components/BasePage';
import AppPaper from '../../../../components/AppPaper';
import AutocompleteInput from '../../../../components/AutocompleteInput';
import { genesViewStyles } from './styles';
import H3 from '../../../../components/typography/H3';
import {
  useGetMostSearchHistoryQuery,
  useGetRecentSearchHistoryQuery,
} from '../../../../services/genesApi/genesApi';
import { useAutocompleteQuery } from '../../hooks/useAutocompleteQuery';
import { useDebounce } from '../../../../hooks/useDebounce';
import GeneNotSelectedContent from './GeneNotSelectedContent';
import GeneSelectedContent from './GeneSelectedContent';
import { IGene } from '../../../../services/genesApi/models';
import { formatMostSearch, formatRecentlySearch } from './utils';

const GenesView = () => {
  const [searchParams] = useSearchParams();

  const defaultGene: IGene = {
    geneDbCrossrefId: Number(searchParams.get('geneDbCrossrefId')),
    geneName: searchParams.get('geneName'),
    locusTag: null,
    id: Number(searchParams.get('geneDbCrossrefId')),
    endPos: 1,
    startPos: 0,
    strand: -1,
  };

  const [inputValue, setInputValue] = useState('');
  const [
    selectedValue,
    setSelectedValue,
  ] = useState<IGene | null>(defaultGene.geneDbCrossrefId ? defaultGene : null);
  const debouncedValue = useDebounce<string>(inputValue, 500);

  const {
    data: mostSearch = [],
    isLoading: isMostSearchLoading, refetch: mostSearchHistoryRefetch,
  } = useGetMostSearchHistoryQuery(undefined, { refetchOnMountOrArgChange: true });

  const {
    data: recentSearch = [],
    isLoading: isRecentSearchLoading, refetch: resentSearchRefetch,
  } = useGetRecentSearchHistoryQuery(undefined, { refetchOnMountOrArgChange: true });

  const {
    data: genes = [], fetchData, isLoading,
  } = useAutocompleteQuery();

  const getOptions = () => {
    const inputEmpty = !inputValue;
    const mostSearchedData = mostSearch
      .slice(-10)
      .sort((prev, next) => next.counter - prev.counter);

    return inputEmpty ? mostSearchedData : genes;
  };

  const options = getOptions();

  const recentSearchLocal = formatRecentlySearch(recentSearch);
  const mostSearchLocal = formatMostSearch(mostSearch);

  const handleInputChange = (e: SyntheticEvent, value: string) => {
    setInputValue(value);
  };

  useEffect(() => {
    if (debouncedValue !== '') {
      fetchData({ search: debouncedValue });
    }
  }, [debouncedValue]);

  const patchedSelectedGene = genes.length === 1
    ? { ...selectedValue, ...genes[0] }
    : selectedValue;

  return (
    <BasePage pageHeader="Gene View">
      <AppPaper>
        <div css={genesViewStyles.autoCompleteContainer}>
          <AutocompleteInput
            isLoading={isLoading}
            onSelectionChanged={(e, value) => {
              if (value === null) {
                mostSearchHistoryRefetch();
                resentSearchRefetch();
              }

              setSelectedValue(value);
            }}
            selectedValue={selectedValue}
            onChange={handleInputChange}
            value={inputValue}
            options={options}
            style={genesViewStyles.sxAutoComplete}
          />
          <H3
            style={genesViewStyles.autoCompleteInputLabel}
          >
            Search Genes by Gene Symbol, NCBI ID or Locus Tag
          </H3>
        </div>
      </AppPaper>
      {
        !selectedValue
          ? (
            <GeneNotSelectedContent
              isMostSearchLoading={isMostSearchLoading}
              isRecentSearchLoading={isRecentSearchLoading}
              mostSearch={mostSearchLocal}
              recentSearch={recentSearchLocal}
              onGeneClick={(item) => {
                setSelectedValue({
                  endPos: 0,
                  startPos: 0,
                  strand: -1,
                  id: item.id,
                  geneDbCrossrefId: item.id,
                  geneName: item.leftColValue as string,
                  locusTag: item.leftColValue as string,
                });
              }}
            />
          )
          : <GeneSelectedContent selectedGene={(patchedSelectedGene) as IGene} />
      }
    </BasePage>
  );
};

export default GenesView;
