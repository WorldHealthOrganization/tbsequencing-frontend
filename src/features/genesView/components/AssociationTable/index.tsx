import AssociationTable from './AssociationTable';

export { LeftColumnButton } from './AssociationTable';
export default AssociationTable;
