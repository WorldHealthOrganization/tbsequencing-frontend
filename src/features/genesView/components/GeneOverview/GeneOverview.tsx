/** @jsxImportSource @emotion/react */
import React from 'react';
import AppPaper from '../../../../components/AppPaper';
import { associationTableStyles } from '../AssociationTable/styles';
import H3 from '../../../../components/typography/H3';
import { IGenesOverviewTableData } from '../../models';
import LoadingWrapper from '../../../../components/LoadingWrapper';

interface IGeneOverviewProps {
  data: IGenesOverviewTableData[];
  tableHeader?: string;
  isLoading: boolean
}

interface IGeneOverviewContentProps {
  data: IGenesOverviewTableData[]
}

const GeneOverviewContent = ({ data }: IGeneOverviewContentProps) => {
  if (!data.length) return <div>No data</div>;
  return (
    <>
      {data.map(({ title, value, element }) => {
        if (value) {
          return (
            <div key={title} css={associationTableStyles.singlePair}>
              <H3 style={associationTableStyles.defaultColumn}>{title}</H3>
              <H3 style={associationTableStyles.rightColumnText}>{element || value}</H3>
            </div>
          );
        }
        return null;
      })}
    </>
  );
};

const GeneOverview = ({ data, tableHeader, isLoading }: IGeneOverviewProps) => (
  <AppPaper style={associationTableStyles.paper}>
    <div>
      {tableHeader && <H3>{tableHeader}</H3>}
    </div>
    <LoadingWrapper isLoading={isLoading} centered>
      <GeneOverviewContent data={data} />
    </LoadingWrapper>
  </AppPaper>
);

export default GeneOverview;
