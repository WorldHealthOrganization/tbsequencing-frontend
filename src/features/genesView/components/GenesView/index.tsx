import GenesView from './GenesView';

export default GenesView;
