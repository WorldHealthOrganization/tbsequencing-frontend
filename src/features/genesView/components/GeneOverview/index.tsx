import GeneOverview from './GeneOverview';

export default GeneOverview;
