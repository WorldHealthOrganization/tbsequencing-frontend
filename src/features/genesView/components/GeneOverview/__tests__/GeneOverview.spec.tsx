import React from 'react';
import { getTreeJSONWithRedux, render } from '../../../../../utils/testUtils/jestUtils';
import { simpleData } from '../../AssociationTable/mock';
import GeneOverview from '../index';
import { formatGenesOverviewData } from '../../GenesView/utils';

const data = formatGenesOverviewData(simpleData);
describe('GeneOverview', () => {
  test('renders GeneOverview component without error', () => {
    render(
      <GeneOverview
        isLoading={false}
        data={data}
      />,
    );
  });

  test('Should render correctly', () => {
    const tree = getTreeJSONWithRedux(
      <GeneOverview
        isLoading={false}
        data={data}
      />,
    );

    expect(tree).toMatchSnapshot();
  });
});
