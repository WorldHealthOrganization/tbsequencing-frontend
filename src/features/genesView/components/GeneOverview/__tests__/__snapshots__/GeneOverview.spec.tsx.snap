// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GeneOverview Should render correctly 1`] = `
<div
  className="css-9j8ib4-AppPaper"
>
  <div />
  <div
    className="css-1w9istt-GeneOverviewContent"
  >
    <p
      className="MuiTypography-root MuiTypography-body1 css-1h9ij6s-MuiTypography-root-BaseText"
    >
      Gene Symbol
    </p>
    <p
      className="MuiTypography-root MuiTypography-body1 css-1h9ij6s-MuiTypography-root-BaseText"
    >
      rpoB
    </p>
  </div>
  <div
    className="css-1w9istt-GeneOverviewContent"
  >
    <p
      className="MuiTypography-root MuiTypography-body1 css-1h9ij6s-MuiTypography-root-BaseText"
    >
      NCBI Gene ID
    </p>
    <p
      className="MuiTypography-root MuiTypography-body1 css-1h9ij6s-MuiTypography-root-BaseText"
    >
      <a
        className="css-1grqmfe"
        href="https://www.ncbi.nlm.nih.gov/gene/888164"
        rel="noreferrer"
        target="_blank"
      >
        888164
      </a>
    </p>
  </div>
  <div
    className="css-1w9istt-GeneOverviewContent"
  >
    <p
      className="MuiTypography-root MuiTypography-body1 css-1h9ij6s-MuiTypography-root-BaseText"
    >
      Gene Description
    </p>
    <p
      className="MuiTypography-root MuiTypography-body1 css-1h9ij6s-MuiTypography-root-BaseText"
    >
      DNA-directed RNA polymerase subunit beta
    </p>
  </div>
  <div
    className="css-1w9istt-GeneOverviewContent"
  >
    <p
      className="MuiTypography-root MuiTypography-body1 css-1h9ij6s-MuiTypography-root-BaseText"
    >
      Locus Tag
    </p>
    <p
      className="MuiTypography-root MuiTypography-body1 css-1h9ij6s-MuiTypography-root-BaseText"
    >
      Rv0667
    </p>
  </div>
  <div
    className="css-1w9istt-GeneOverviewContent"
  >
    <p
      className="MuiTypography-root MuiTypography-body1 css-1h9ij6s-MuiTypography-root-BaseText"
    >
      Gene type
    </p>
    <p
      className="MuiTypography-root MuiTypography-body1 css-1h9ij6s-MuiTypography-root-BaseText"
    >
      protein coding
    </p>
  </div>
  <div
    className="css-1w9istt-GeneOverviewContent"
  >
    <p
      className="MuiTypography-root MuiTypography-body1 css-1h9ij6s-MuiTypography-root-BaseText"
    >
      Location
    </p>
    <p
      className="MuiTypography-root MuiTypography-body1 css-1h9ij6s-MuiTypography-root-BaseText"
    >
      759,807..763,325
    </p>
  </div>
  <div
    className="css-1w9istt-GeneOverviewContent"
  >
    <p
      className="MuiTypography-root MuiTypography-body1 css-1h9ij6s-MuiTypography-root-BaseText"
    >
      Length
    </p>
    <p
      className="MuiTypography-root MuiTypography-body1 css-1h9ij6s-MuiTypography-root-BaseText"
    >
      3,518 nt
    </p>
  </div>
  <div
    className="css-1w9istt-GeneOverviewContent"
  >
    <p
      className="MuiTypography-root MuiTypography-body1 css-1h9ij6s-MuiTypography-root-BaseText"
    >
      CDS
    </p>
    <p
      className="MuiTypography-root MuiTypography-body1 css-1h9ij6s-MuiTypography-root-BaseText"
    >
      <a
        className="css-rgkkh9"
        href="https://www.ncbi.nlm.nih.gov/protein/NP_215181.1"
        rel="noreferrer"
        target="_blank"
      >
        NP_215181.1
      </a>
    </p>
  </div>
  <div
    className="css-1w9istt-GeneOverviewContent"
  >
    <p
      className="MuiTypography-root MuiTypography-body1 css-1h9ij6s-MuiTypography-root-BaseText"
    >
      Protein Length
    </p>
    <p
      className="MuiTypography-root MuiTypography-body1 css-1h9ij6s-MuiTypography-root-BaseText"
    >
      1,172 aa
    </p>
  </div>
</div>
`;
